lib/src/discovery_analysis.dart

```diff
@@ -0,0 +1,176 @@
+// Copyright (c) 2015, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// Analyse a directory structure and find packages resolvers for each
+/// subdirectory.
+///
+/// The resolvers are generally the same that would be found by using
+/// the `discovery.dart` library on each subdirectory in turn,
+/// but more efficiently.
+library package_config.discovery_analysis;
+
+import "dart:collection" show HashMap;
+import "dart:io" show Directory, Platform;
+
+import "package:path/path.dart" as path;
+
+import "package_config.dart";
+import "discovery.dart";
+
+/// Associates a [PackageConfig] with a directory.
+///
+/// The package resolution applies to the directory and any subdirectory
+/// that doesn't have its own overriding child [PackageContext].
+abstract class PackageContext {
+  /// The directory which introduced the [packageCOnfig] resolver.
+  Directory get directory;
+
+  /// The [PackageConfig] that applies to the directory.
+  ///
+  /// Introduced either by a `.dart_tool/package_config.json` file or a
+  /// `.packages` file in the [directory].
+  PackageConfig get packageConfig;
+
+  /// Look up the [PackageConfig] that applies to a specific directory.
+  ///
+  /// The [subdirectory] must be inside [directory].
+  PackageConfig operator [](Directory subdirectory);
+
+  /// Child immediate contexts that apply to subdirectories of [directory].
+  Iterable<PackageContext> get children;
+
+  /// A map from directory to package resolver.
+  ///
+  /// Has an entry for this package context and for each immediate child context
+  /// contained in this one.
+  Map<Directory, PackageConfig> asMap();
+
+  /// Analyze [directory] and subdirectories for package configurations.
+  ///
+  /// Returns a context mapping subdirectories to [PackageConfig] objects.
+  ///
+  /// The analysis assumes that there are no package configuration files
+  /// in a parent directory of `directory`.
+  /// If there is, its corresponding [PackageConfig] object
+  /// should be provided as [root].
+  ///
+  /// Each directory visited is checked for the presence of a `.packages`
+  /// or `.dart_tool/package_config.json` file. If such a file exists,
+  /// it is loaded. If loading fails, the directory is treated as if
+  /// no configuration was found.
+  /// If [onError] is supplied, the error that caused loading to fail
+  /// is reported to that function, otherwise it's ignored.
```

Comment:
Because it might be a false negative. If a directory happens to contain a file named `.packages` for any other reason, but it's not a package resolution file, then we ignore it instead of making it an error.  

Even if it's a true positive, the scope of `PackageContext` is larger than the individual file. Consider if this functionality was used by the analyzer (which was the original reason to add it). Then I don't want the analyzer to give up on analyzing entirely just because one sub-directory has an error.

The `onError` is intended for allowing the tool to show diagnostic reports, like `foo/bar/baz/.packags is not a valid package configuration file`, rather than giving up entirely.